### DISCUSSION TITLE
Logging Object Detection Results of Video in the Excel File

### DISCUSSION THREAD

**Discussion by Aar-Kay** [ultralytics/yolov5#2133] [Q&A] [answered]
Hi,

I am new to coding and posting first time on GitHub.

Well, I was able to successfully train the model on the custom dataset and I am using it to detect some signs in the videos.
I can see the results in the video as well as in the terminal.

Now I want my results to be saved in an excel file which should have a log of all the objects detected **along with the time of the video**.

![image](https://user-images.githubusercontent.com/62192088/106952198-13e89480-6731-11eb-980e-409320442a89.png)

Can someone please suggest how it can be done?

**Comment by glenn-jocher** [comment, marked as answer]
@Aar-Kay hey buddy, congrats on your first training! You can produce detection logs as text files from video frames like this:
```bash
python detect.py --source video.mp4 --save-txt
```

The part of detect.py that saves these text files is here, you could introduce your own code to update an excel file instead:
https://github.com/ultralytics/yolov5/blob/73a066993051339f6adfe5095a7852a2b9184c16/detect.py#L103-L110

**Reply by Aar-Kay** [reply]
@glenn-jocher Thank you so much. I will work on it.

**Reply by mintu07ruet** [reply]
Aar-Kay @ Could you please share your code for time stamping and excel output. I am trying to do the same for bicycle only. Thanks!

**Reply by Aar-Kay** [reply]
@mintu07ruet  I made some changes in the detect.py file

Below is the final code that I used for my project

import argparse
import time
from csv import DictWriter
from pathlib import Path

import cv2
import torch
import torch.backends.cudnn as cudnn
from numpy import random

from models.experimental import attempt_load
from utils.datasets import LoadStreams, LoadImages
from utils.general import check_img_size, non_max_suppression, apply_classifier, scale_coords, xyxy2xywh, \
    strip_optimizer, set_logging, increment_path
from utils.plots import plot_one_box
from utils.torch_utils import select_device, load_classifier, time_synchronized


def detect(save_img=False):
    source, weights, view_img, save_txt, imgsz = opt.source, opt.weights, opt.view_img, opt.save_txt, opt.img_size
    webcam = source.isnumeric() or source.endswith('.txt') or source.lower().startswith(
        ('rtsp://', 'rtmp://', 'http://'))

    # Directories
    save_dir = Path(increment_path(Path(opt.project) / opt.name, exist_ok=opt.exist_ok))  # increment run
    (save_dir / 'labels' if save_txt else save_dir).mkdir(parents=True, exist_ok=True)  # make dir

    # Initialize
    set_logging()
    device = select_device(opt.device)
    half = device.type != 'cpu'  # half precision only supported on CUDA

    # Load model
    model = attempt_load(weights, map_location=device)  # load FP32 model
    imgsz = check_img_size(imgsz, s=model.stride.max())  # check img_size
    if half:
        model.half()  # to FP16

    # Second-stage classifier
    classify = False
    if classify:
        modelc = load_classifier(name='resnet101', n=2)  # initialize
        modelc.load_state_dict(torch.load('weights/resnet101.pt', map_location=device)['model']).to(device).eval()

    # Set Dataloader
    vid_path, vid_writer = None, None
    if webcam:
        view_img = True
        cudnn.benchmark = True  # set True to speed up constant image size inference
        dataset = LoadStreams(source, img_size=imgsz)
    else:
        save_img = True
        dataset = LoadImages(source, img_size=imgsz)

    # Get names and colors
    names = model.module.names if hasattr(model, 'module') else model.names
    colors = [[random.randint(0, 255) for _ in range(3)] for _ in names]

    # Run inference
    with open('log_file.csv', 'a') as csvfile:
      header = ('Frame', 'AutoSteer', 'AutoSteer Confidence', 'Indicator', 'Indicator Confidence', 'ProPilot', 'ProPilot Confidence', 'VID - Lanes Detection Status', 'VID LDS Confidence', 'VID - LDW', 'VID LDW Confidence')
      csv_writer = DictWriter(csvfile, fieldnames=header, lineterminator='\n', delimiter=',')
      csv_writer.writeheader()

    t0 = time.time()
    img = torch.zeros((1, 3, imgsz, imgsz), device=device)  # init img
    _ = model(img.half() if half else img) if device.type != 'cpu' else None  # run once
    
    for path, img, im0s, vid_cap in dataset:
        img = torch.from_numpy(img).to(device)
        img = img.half() if half else img.float()  # uint8 to fp16/32
        img /= 255.0  # 0 - 255 to 0.0 - 1.0
        if img.ndimension() == 3:
            img = img.unsqueeze(0)

        # Inference
        t1 = time_synchronized()
        pred = model(img, augment=opt.augment)[0]

        # Apply NMS
        pred = non_max_suppression(pred, opt.conf_thres, opt.iou_thres, classes=opt.classes, agnostic=opt.agnostic_nms)
        t2 = time_synchronized()

        # Apply Classifier
        if classify:
            pred = apply_classifier(pred, modelc, img, im0s)

        # Process detections
        for i, det in enumerate(pred):  # detections per image
            if webcam:  # batch_size >= 1
                p, s, im0, frame = Path(path[i]), '%g: ' % i, im0s[i].copy(), dataset.count
            else:
                p, s, im0, frame = Path(path), '', im0s, getattr(dataset, 'frame', 0)

            save_path = str(save_dir / p.name)
            txt_path = str(save_dir / 'labels' / p.stem) + ('_%g' % dataset.frame if dataset.mode == 'video' else '')
            s += '%gx%g ' % img.shape[2:]  # print string
            gn = torch.tensor(im0.shape)[[1, 0, 1, 0]]  # normalization gain whwh
            if len(det):
                # Rescale boxes from img_size to im0 size
                det[:, :4] = scale_coords(img.shape[2:], det[:, :4], im0.shape).round()

                # Print results
                for c in det[:, -1].unique():
                    n = (det[:, -1] == c).sum()  # detections per class
                    s += '%g %ss, ' % (n, names[int(c)])  # add to string

                # Write results
                tl1 = cnf1 = tl2 = cnf2 = tl3 = cnf3 = tl4 = cnf4 = tl5 = cnf5 = ''
                for *xyxy, conf, cls in reversed(det):
                    if save_img or view_img:  # Add bbox to image
                        label = '%s %.2f' % (names[int(cls)], conf)
                        plot_one_box(xyxy, im0, label=label, color=colors[int(cls)], line_thickness=3)

                    if save_txt:
                      icls = int(cls)
                      if icls == 0 or icls == 1:
                        tl1 = int(cls)
                        cnf1 = '%.2f' % (conf)
                      if icls == 2 or icls == 3:
                        tl2 = int(cls)
                        cnf2 = '%.2f' % (conf)
                      if icls == 4 or icls == 5 or icls == 6:
                        tl3 = int(cls)
                        cnf3 = '%.2f' % (conf)
                      if icls == 7 or icls == 8:
                        tl4 = int(cls)
                        cnf4 = '%.2f' % (conf)
                      if icls == 9 or icls == 10:
                        tl5 = int(cls)
                        cnf5 = '%.2f' % (conf)

                with open('log_file.csv', 'a') as csvfile:
                  csv_writer = DictWriter(csvfile, fieldnames=header, lineterminator='\n', delimiter=',')
                  csv_writer.writerow({'Frame': frame, 'AutoSteer': tl1, 'AutoSteer Confidence': cnf1, 'Indicator': tl2, 'Indicator Confidence': cnf2, 'ProPilot': tl3, 'ProPilot Confidence': cnf3, 'VID - Lanes Detection Status': tl4, 'VID LDS Confidence': cnf4, 'VID - LDW': tl5, 'VID LDW Confidence': cnf5})




            # Print time (inference + NMS)
            print('%sDone. (%.3fs)' % (s, t2 - t1))

            # Stream results
            if view_img:
                cv2.imshow(str(p), im0)
                if cv2.waitKey(1) == ord('q'):  # q to quit
                    raise StopIteration

            # Save results (image with detections)
            if save_img:
                if dataset.mode == 'images':
                    cv2.imwrite(save_path, im0)
                else:
                    if vid_path != save_path:  # new video
                        vid_path = save_path
                        if isinstance(vid_writer, cv2.VideoWriter):
                            vid_writer.release()  # release previous video writer

                        fourcc = 'mp4v'  # output video codec
                        fps = vid_cap.get(cv2.CAP_PROP_FPS)
                        print(f'FPS of video {fps}')
                        w = int(vid_cap.get(cv2.CAP_PROP_FRAME_WIDTH))
                        h = int(vid_cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
                        vid_writer = cv2.VideoWriter(save_path, cv2.VideoWriter_fourcc(*fourcc), fps, (w, h))
                    vid_writer.write(im0)

    if save_txt or save_img:
        s = f"\n{len(list(save_dir.glob('labels/*.txt')))} labels saved to {save_dir / 'labels'}" if save_txt else ''
        print(f"Results saved to {save_dir}{s}")

    print('Done. (%.3fs)' % (time.time() - t0))


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--weights', nargs='+', type=str, default='yolov5s.pt', help='model.pt path(s)')
    parser.add_argument('--source', type=str, default='data/images', help='source')  # file/folder, 0 for webcam
    parser.add_argument('--img-size', type=int, default=640, help='inference size (pixels)')
    parser.add_argument('--conf-thres', type=float, default=0.25, help='object confidence threshold')
    parser.add_argument('--iou-thres', type=float, default=0.45, help='IOU threshold for NMS')
    parser.add_argument('--device', default='', help='cuda device, i.e. 0 or 0,1,2,3 or cpu')
    parser.add_argument('--view-img', action='store_true', help='display results')
    parser.add_argument('--save-txt', action='store_true', help='save results to *.txt')
    parser.add_argument('--save-conf', action='store_true', help='save confidences in --save-txt labels')
    parser.add_argument('--classes', nargs='+', type=int, help='filter by class: --class 0, or --class 0 2 3')
    parser.add_argument('--agnostic-nms', action='store_true', help='class-agnostic NMS')
    parser.add_argument('--augment', action='store_true', help='augmented inference')
    parser.add_argument('--update', action='store_true', help='update all models')
    parser.add_argument('--project', default='runs/detect', help='save results to project/name')
    parser.add_argument('--name', default='exp', help='save results to project/name')
    parser.add_argument('--exist-ok', action='store_true', help='existing project/name ok, do not increment')
    opt = parser.parse_args()
    print(opt)

    with torch.no_grad():
        if opt.update:  # update all models (to fix SourceChangeWarning)
            for opt.weights in ['yolov5s.pt', 'yolov5m.pt', 'yolov5l.pt', 'yolov5x.pt']:
                detect()
                strip_optimizer(opt.weights)
        else:
            detect()

**Reply by mintu07ruet** [reply]
Thank you so much @Aar-Kay  for sharing the code. I will go through in it. Appreciate!

**Reply by mintu07ruet** [reply]
Hi Aar-Kay
I am getting the following error
ImportError: cannot import name 'plot_one_box' from 'utils.plots' (C:\Users\mxm5116\yolov5\utils\plots.py)
 any idea what can solve this issue? Also may I get your email, so that I can contact with through email. Thanks!

**Reply by mintu07ruet** [reply]
Hi Aar-kay,
There was mismatch with version which produced the error. your modified detect.py file is running, but could you please mention, how can I get the time stamp as in the log file, it is recording Frame, but I did not see any time stamp. Thanks

**Reply by renjirow** [reply]
can u share to us the file of ur project

**Reply by Asphyxcome** [reply]
hi, thanks for sharing the code, but when I tried to run the code, I got a few errors. Could you send the whole file where you made the changes and tell me how to output the data from yolo to excel file?

**Comment by mintu07ruet** [comment]
glenn-jocher@ can detect.py file can read video file directly from outlook one drive?

**Reply by glenn-jocher** [reply]
@mintu07ruet yes detect.py can load remote URL sources, i.e.:

```python
!python detect.py --weights yolov5s.pt --source https://ultralytics.com/assets/decelera_landscape.mov  # video URL
```

**Reply by glenn-jocher** [reply]
<img width="1206" alt="Screenshot 2021-11-05 at 17 50 40" src="https://user-images.githubusercontent.com/26833433/140548029-9eca0b50-837e-471e-b3a4-dc8939b8f319.png">

**Reply by mintu07ruet** [reply]
![image](https://user-images.githubusercontent.com/60245323/140549050-bbdb7a1b-b09f-443e-9a1a-e8a52d9c3955.png)
glenn-jocher@ I am getting the above error when I am trying to read the video file from outlook onedrive. Any suggestions to run it. TIA!

**Reply by glenn-jocher** [reply]
@mintu07ruet if you have a complex URL you should always enclose it in quotation marks `--source 'URL'`

**Reply by parmarth-1208** [reply]
hey i just wanted to ask in the below code what path will replace "yolov8.weights" and "yolov8.cfg" ?
import cv2
import numpy as np
from itertools import combinations
import openpyxl

# Function to calculate center point of a bounding box
def calculate_center(bbox):
    return ((bbox[0] + bbox[2]) / 2, (bbox[1] + bbox[3]) / 2)

# Function to compute Euclidean distance between two points
def euclidean_distance(point1, point2):
    return np.sqrt((point1[0] - point2[0])**2 + (point1[1] - point2[1])**2)

# Function to group characters based on proximity
def group_characters(bboxes):
    center_points = [calculate_center(bbox) for bbox in bboxes]
    distance_matrix = np.zeros((len(bboxes), len(bboxes)))
    for i, center1 in enumerate(center_points):
        for j, center2 in enumerate(center_points):
            distance_matrix[i, j] = euclidean_distance(center1, center2)

    proximity_threshold = 50
    groups = []
    for i, j in combinations(range(len(bboxes)), 2):
        if distance_matrix[i, j] < proximity_threshold:
            group_found = False
            for group in groups:
                if i in group or j in group:
                    group.add(i)
                    group.add(j)
                    group_found = True
                    break
            if not group_found:
                groups.append({i, j})
    return groups

# Load the pre-trained YOLOv8 model
net = cv2.dnn.readNet("yolov8.weights", "yolov8.cfg")
classes = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
           "A", "B", "C", "D", "E", "F", "G", "H", "I", "J",
           "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T",
           "U", "V", "W", "X", "Y", "Z"]
output_layers = net.getUnconnectedOutLayersNames()

# Create Excel workbook and worksheet
wb = openpyxl.Workbook()
ws = wb.active
ws.append(["Plate Number"])

# Process video
video_capture = cv2.VideoCapture("your_video.mp4")
while True:
    ret, frame = video_capture.read()
    if not ret:
        break

    # Perform object detection on the frame
    blob = cv2.dnn.blobFromImage(frame, 0.00392, (416, 416), (0, 0, 0), True, crop=False)
    net.setInput(blob)
    outs = net.forward(output_layers)

    # Extract bounding boxes and filter character detections
    bboxes = []
    for out in outs:
        for detection in out:
            scores = detection[5:]
            class_id = np.argmax(scores)
            confidence = scores[class_id]
            if confidence > 0.5:  # Confidence threshold
                center_x = int(detection[0] * frame.shape[1])
                center_y = int(detection[1] * frame.shape[0])
                w = int(detection[2] * frame.shape[1])
                h = int(detection[3] * frame.shape[0])
                x = int(center_x - w / 2)
                y = int(center_y - h / 2)
                if classes[class_id] in classes:
                    bboxes.append((x, y, x + w, y + h))

    # Group characters
    groups = group_characters(bboxes)

    # Extract plate numbers and store in Excel worksheet
    for group in groups:
        plate_number = ''.join([classes[idx] for idx in group])
        ws.append([plate_number])

    # Draw bounding boxes and display
    for bbox in bboxes:
        cv2.rectangle(frame, (bbox[0], bbox[1]), (bbox[2], bbox[3]), (0, 255, 0), 2)
    cv2.imshow("ANPR", frame)
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# Save Excel workbook
wb.save("plate_numbers.xlsx")

video_capture.release()
cv2.destroyAllWindows()

**Comment by mintu07ruet** [comment]
glenn-jocher@ you are right! the url is too complex. I also used to '  ' but did not work, same error.
![image](https://user-images.githubusercontent.com/60245323/140551850-f7f99d54-2c8e-4c9f-81e2-c3b840098242.png)
 Here is my video link: https://mavsuta-my.sharepoint.com/:v:/r/personal/nrgomez_uta_edu/Documents/Mdmintu%20Video%20Experiment/Exp_2021630_12_55_2.avi?csf=1&web=1&e=kgbjLo/Exp_2021630_12_55_2.avi

if you want to test! Thanks for the help!

**Reply by glenn-jocher** [reply]
@mintu07ruet ah I see. In your case the URL is not publically available for download. If you click your link in a browser a download does not start, it is hidden behind a login screen, so naturally this video will not work. detect.py will work with any publically available image or video URL like https://ultralytics.com/assets/decelera_landscape.mov

**Reply by glenn-jocher** [reply]
@mintu07ruet I just pushed #5525 to improve URL inference. The following two examples will work well now, i.e. you can test them in our Colab notebook:
https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb
```python
!python detect.py --weights yolov5s.pt --source https://ultralytics.com/assets/zidane.jpg  # image URL
!python detect.py --weights yolov5s.pt --source https://ultralytics.com/assets/decelera_landscape.mov  # video URL
```

**Reply by mintu07ruet** [reply]
Thank you gleen-jocher@ for all the explanation. I will check to see if we can make our project video data public. But it is confidential in nature and big amount of video data like thousands of hour data that need to evaluate.  We dumped all the video data in one drive which can share a particular person, may be not  publicly. We prefer to run yolov5 in our local machine to save the output quickly. That's why we are not using kaggle and colab due to limitation of their GPU use. 

If you have any updates on how to run the yolov5s.pt to run shared video from one drive directly in future. Please update me. Thank you so much for your time, help and suggestions! will back if I face further problems! Thanks!

**Reply by glenn-jocher** [reply]
@mintu07ruet we don't have the time to view your custom data, but in general you should run inference and training on local data for fastest results and lowest deployment risk.

**Comment by saktheeswaranswan** [comment]
> Hi,
> 
> I am new to coding and posting first time on GitHub.
> 
> Well, I was able to successfully train the model on the custom dataset and I am using it to detect some signs in the videos.
> I can see the results in the video as well as in the terminal.
> 
> Now I want my results to be saved in an excel file which should have a log of all the objects detected **along with the time of the video**.
> 
> ![image](https://user-images.githubusercontent.com/62192088/106952198-13e89480-6731-11eb-980e-409320442a89.png)
> 
> Can someone please suggest how it can be done?   
> 
Here is all you want with the code its in my you tube channel bi have made an video on how to save yolo object detection with time stamp try to understand it the git link is given below the video description
https://youtu.be/1pw18Ml_SQ4

**Reply by saktheeswaranswan** [reply]
Here is the another code with some error correction and the code works good the if you do it on pandas it will work good and in Linux it gives permission to excel file but in windows ir shows permission error 13 which I have tried my best to rectify so check the links below first one shows about the code and the second one gives the permission error 13 in excel if it comes how to rectify it there is also another repo which uses pandas and csv with object detection tracking I'd with vehicle speed and time one of my subscriber told me that yolo v5 .pt pytorch weights also works on that which Is given in the video description so check it

https://youtu.be/360xRAnk1O0
Permission  error 13 in excel which comes 8n windows give it acess and if it shows up move to ubantu Linux and run the code that's all I can do please support me on youtube by subscribing 
https://youtu.be/fbsP4Az-LyQ

My final advice would be get administrator acess to Microsoft excel or use open office excel even if it shows error give full rights and run it error 13 permission denied for excel will not show up I guess this is my small contribution to our open source community

**Reply by saktheeswaranswan** [reply]
happy news problem solved  but subscribe to  random neural monk  YouTube channel  this works without showing perrmission 13 error so enjoy to log csv with time stamp class id also made with chat gpt took 6 hours

https://github.com/saktheeswaranswan/yolov5-csv-yolov5excel-crop-needs-yolov5-rasberrypi-GPIO

**Comment by Emmanuel-Muroboka** [comment]
Hello Mr @ glenn-jocher i have a question that am in need of your help with, is it possible to get the object detected information, say the names of a detected class, and display it on a different platform using the OPC-UA system? I have two computers one is running object detection which is our OPC client then the other computer is running the OPC server. I have created a node on the server which then the client subscribes to and can change the values of the node, so what I would want to show is the results am obtaining from the object detection to show on the server. How can I achieve this?
